### PR TITLE
Generate merge spec

### PR DESCRIPTION
**Pull request by borsch**
Implementation for https://github.com/OpenAPITools/openapi-generator/issues/3981

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by borsch** [comment]
[@wing328](https://github.com/wing328) (2015/07) [❤️](https://www.patreon.com/wing328)
    [@jimschubert](https://github.com/jimschubert) (2016/05) [❤️](https://www.patreon.com/jimschubert)
    [@cbornet](https://github.com/cbornet) (2016/05)
    [@jmini](https://github.com/jmini) (2018/04) [❤️](https://www.patreon.com/jmini)
    [@etherealjoy](https://github.com/etherealjoy) (2019/06)
    [@spacether](https://github.com/spacether) (2020/05) [❤️](https://github.com/sponsors/spacether/)

**Comment by borsch** [comment]
@wing328 could you please review once you have a time?

**Comment by wing328** [comment]
@borsch thanks for the PR as always.

do we need to update openapi-generator-cli module as well?

currently, the users need to use a script or something to loop through all the spec in a folder in order to achieve the same ,right?

**Comment by borsch** [comment]
> @borsch thanks for the PR as always.
> 
> do we need to update openapi-generator-cli module to support the new option as well?
> 
> currently, the users need to use a script or something to loop through all the spec in a folder in order to achieve the same ,right?

Thanks. Sure, will add CLI support. 
Regarding workaround - personally I use small bash script to handle this, but this is a little bit annoying to move it to each and every service, where generator is used


```bash
#!/bin/bash
API_HEADER="openapi: '3.0.0'
info:
    description: Merge description
    version: '1.0'
    title: Merge title

servers:
  - url: https://localhst:8079

paths:\n"
printf "%s${API_HEADER}" >> "src/main/resources/contract/_index_auto.yaml"
grep --exclude='*index*' -E '^ */' "src/main/resources/contract"/* | sort | while read -r line; do
  path=$(echo "$line" | awk '{print $2}')
  fileName=$(echo "$line" | awk '{print $1}' | awk -F "/" '{print $NF}' | sed 's/:*$//g')
  # shellcheck disable=SC2001
  ref=$(echo "${path//\//~1}" | sed 's/:*$//g')

  printf "
  %s$path
    \$ref: './%s$fileName#/paths/%s$ref'\n" >> "src/main/resources/contract/_index_auto.yaml"
done
```

**Comment by borsch** [comment]
@wing328 I've applied change to CLI module

**Comment by wing328** [comment]
i just merged a PR with updated samples. can you merge the latest master into your branch?

**Comment by wing328** [comment]
Did some tests and the result is good 👍 

I will merge when all CI tests pass. 

Thanks again for the PR.

**Comment by heyline** [comment]
Hello, I encountered several problems when using `inputSpecRootDirectory` and would like to ask for advice.

1. `inputSpecRootDirectory` attribute generated merged.yaml for me, can I customize this name?

2. If I have a secondary directory, like `openapi/path/configuration.yml` , I find that merge.yml generates the wrong path, the correct path is
```
  /connectors:
    $ref: "./path/configuration.yml#/paths/~1configuration"
```
but i got
```
  /configuration:
    $ref: "./path\\configuration.yml#/paths/~1configuration"
```

3. The generated file does not contain my tags.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/config/MergedSpecBuilder.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    private static Map<String, Object> generateHeader(String openapiVersion) {
+        Map<String, Object> map = new HashMap<>();
+        map.put("openapi", openapiVersion);
+        map.put("info", ImmutableMap.of(
+            "title", "merged spec",
+            "description", "merged spec",
+            "version", "1.0.0"
+        ));
+        map.put("servers", Collections.singleton(
+            ImmutableMap.of("url", "http://localhost:8080")
```
[the remark this answers is not in the document]
Comment (heyline) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/config/MergedSpecBuilder.java:116]: I have abandoned this plugin and use redocly

```
tasks.register<Exec>("mergeOpenApiSpecs") {
    val specs = projectDir.resolve("src/main/resources/openapi").listFiles()!!.filter { it.name != "merged.yml" }
        .map { it.name }.toTypedArray()
    commandLine(
        "docker",
        "run",
        "--rm",
        "-v",
        "${projectDir.resolve("src/main/resources/openapi").absolutePath}:/spec",
        "redocly/cli:1.12.2",
        "join",
        "--without-x-tag-groups",
        *specs,
        "-o",
        "merged.yml"
    )
}
```

[the remark this answers is not in the document]
Comment (gd-scalemem) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/config/MergedSpecBuilder.java:116]: Doing the merge explicitly is not that hard and offers more flexibility, but would be good to have an built-in one that does the job for many use cases. @borsch

**Pull request merged**